hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/operations/PartitionIteratingOperation.java

```diff
@@ -181,19 +173,52 @@ protected void toString(StringBuilder sb) {
         sb.append(", operationFactory=").append(operationFactory);
     }
 
-    private static class ResponseQueue implements OperationResponseHandler {
+    private class OperationResponseHandlerImpl implements OperationResponseHandler {
 
-        private final BlockingQueue<Object> queue = ResponseQueueFactory.newResponseQueue();
+        // an array with the partitionCount as length, so we can quickly do a lookup for a given partitionId.
+        // it will store all the 'sub' responses.
+        private final AtomicReferenceArray<Object> responseArray = new AtomicReferenceArray<Object>(
+                getNodeEngine().getPartitionService().getPartitionCount());
+
+        // contains the number of pending operations. If it hits zero, all responses have been received.
+        private final AtomicInteger pendingOperations;
+        private final int[] partitions;
+
+        OperationResponseHandlerImpl(int[] partitions) {
+            this.partitions = partitions;
+            this.pendingOperations = new AtomicInteger(partitions.length);
+        }
 
         @Override
-        public void sendResponse(Operation op, Object obj) {
-            if (!queue.offer(obj)) {
-                throw new HazelcastException("Response could not be queued for transportation");
+        public void sendResponse(Operation op, Object response) {
+            if (response instanceof NormalResponse) {
+                response = ((NormalResponse) response).getValue();
+            } else if (response == null) {
+                response = NULL;
+            }
+
+            // we try to set the response in the responseArray
+            if (!responseArray.compareAndSet(op.getPartitionId(), null, response)) {
+                //duplicate response; should not happen.
```

Comment:
I think it is a good idea to add some detection.

What about logging because this method is going to be executed from some general purpose executor which will deal with the assert using logging. And now we can provide a bit more context. Also it will log when assertions are disabled.